File: 2022/lua/2022/day01/day01.lua
#!/usr/bin/env lua
-- https://adventofcode.com/2022/day/1

if #arg ~= 1 then
  io.stderr:write('Usage: ./day02.lua input.txt')
  os.exit(1)
end

local function part1(file)
  local max = function(t)
    local mx = t[1]
    for i=2,#t do
      if mx < t[i] then mx = t[i] end
    end
    return mx
  end
  local sum,ret = 0,{}
  for line in io.lines(file) do
    if not line:match('^$') then
      sum = sum + tonumber(line)
    else
      ret[#ret+1] = sum
      sum = 0
    end
  end
  return max(ret)
end

local function part2(file)
  local sum,ret = 0,{}
  for line in io.lines(file) do
    if not line:match('^$') then
      sum = sum + tonumber(line)
    else
      ret[#ret+1] = sum
      sum = 0
    end
  end
  table.sort(ret,function(a,b) return a > b end)
  return ret[1] + ret[2] + ret[3]
end

print(string.format("Part1: %d", part1(arg[1])))
print(string.format("Part2: %d", part2(arg[1])))

File: 2022/perl/day01/day01.pl
#!/usr/bin/env perl
# https://adventofcode.com/2022/day/1
use strict;
use warnings;
use List::Util qw(max sum);

die "Usage: $0 input.txt" unless @ARGV == 1
  && $ARGV[0] =~ /\w+/;

sub part1{
  my ($file) = @_;
  open(my $fh,'<',$file) or die "Can't open file: $!";
  my ($sum,@ret) = (0);
  while(<$fh>){
    $sum += $_ unless(/^$/);
    do{push @ret,$sum; $sum = 0} if(/^$/);
  }
  close($fh);
  max @ret;
}

sub part2{
  my ($file) = @_;
  open(my $fh,'<',$file) or die "Can't open file: $!";
  my ($sum,@ret) = (0);
  while(<$fh>){
    $sum += $_ unless(/^$/);
    do{push @ret,$sum; $sum = 0} if(/^$/);
  }
  close($fh);
  @ret = sort {$b <=> $a} @ret;
  sum @ret[0..2];
}

printf "Part1: %d\n", part1($ARGV[0]);
printf "Part2: %d\n", part2($ARGV[0]);

